### PR TITLE
opt: improve the performance of sending data

### PR DESCRIPTION
**Pull request by limpo1989**
When data reading occurs, there is a high probability of data writeback generated in the `OnTraffic`.
At this time, it is not necessary to wait for the registration write event callback to immediately perform the write operation, which can improve the performance of sending data.

benchmark use https://github.com/gnet-io/gnet-benchmarks

### BEFORE
```
Using interface lo to connect to [127.0.0.1]:7002
Ramped up to 500 connections.
Total data sent:     35112.2 MiB (36817855642 bytes)
Total data received: 14540.8 MiB (15247174323 bytes)
Bandwidth per channel: 13883.816⇅ Mbps (1735477.0 kBps)
Aggregate bandwidth: 4065.857↓, 9817.959↑ Mbps
Packet rate estimate: 372404.8↓, 858515.8↑ (12↓, 41↑ TCP MSS/op)
Test duration: 30.0004 s.
```

### AFTER
```
Using interface lo to connect to [127.0.0.1]:7002
Ramped up to 500 connections.
Total data sent:     41217.8 MiB (43220020514 bytes)
Total data received: 27600.4 MiB (28941133491 bytes)
Bandwidth per channel: 19242.419⇅ Mbps (2405302.3 kBps)
Aggregate bandwidth: 7717.413↓, 11525.006↑ Mbps
Packet rate estimate: 706735.2↓, 999879.1↑ (12↓, 38↑ TCP MSS/op)
Test duration: 30.0009 s.
```

### RESULT

before:
`Aggregate bandwidth: 4065.857↓, 9817.959↑ Mbps`   
after:  
 `Aggregate bandwidth: 7717.413↓, 11525.006↑ Mbps`

### CODE REVIEW

File: connection_bsd.go [hunk 1]
```diff
@@ -54,5 +57,13 @@ func (c *conn) processIO(_ int, filter netpoll.IOEvent, flags netpoll.IOFlags) (
 			err = el.close(c, io.EOF)
 		}
 	}
+
+	// When data reading occurs, there is a high probability of data writeback generated in the `OnTraffic`.
+	// At this time, it is not necessary to wait for the registration write event callback to immediately perform the write operation,
+	// which can improve the performance of sending data.
+	if haveRead && nil == err && c.OutboundBuffered() > 0 {
```
Comment (panjf2000) [inline connection_bsd.go:64]: `haveRead` 这个变量没什么必要，直接判断 `filter == unix.EVFILT_READ` 就行了。

**Comment by panjf2000** [comment]
On what platform did you run the benchmark? Also which version of gnet you were using for benchmark?

File: connection_bsd.go (hunk 1 above)
Comment (panjf2000) [reply connection_bsd.go:64]: 这里也有一个风险点：对端现在不可写，那这里就多浪费一次系统调用。暂时不知道这个风险会对总体效果造成多大的影响，比如一些特殊应用场景的业务，如果是发送快而接收慢，导致 socket buffer 里总有堆积，那服务端这种模式可能就会浪费很多系统调用。。。。需要在权衡一下。

**Comment by limpo1989** [comment]
```
 OS: Ubuntu 20.04.6 LTS on Windows 10 x86_64
 Kernel: 5.15.146.1-microsoft-standard-WSL2
 Uptime: 1 day, 12 hours, 8 mins
 Packages: 846 (dpkg), 4 (snap)
 Shell: bash 5.0.17
 Theme: Adwaita [GTK3]
 Icons: Adwaita [GTK3]
 Terminal: Relay(1325)
 CPU: AMD Ryzen 9 7945HX with Radeon Graphics (32) @ 2.495GHz
 GPU: 2060:00:00.0 Microsoft Corporation Device 008e
 Memory: 498MiB / 31951MiB
```

benchmark version use `v2.5.2`

This optimization may not be as good as you think. If there is data to send immediately after OnTraffic processing, it will usually go directly into the socket buffer and not into the outbound buffer. Unless the socket is not writable, this change may introduce invalid write operations and cause more CPU consumption.

Perhaps this change will improve intranet services, such as the rpc framework serving intranets.

File: connection_bsd.go (hunk 1 above)
Comment (limpo1989) [reply connection_bsd.go:64]: 是的，也许应该加一个开关，内网服务应该还是有收益的，比如服务于内网的rpc框架，通常内网不存在堆积，比如这个压测结果就是典型内网场景

**Comment by limpo1989** [comment]
Write operations are processed first during epoll-wait. If outbound buffer is still not cleared, it means that socket has entered an unwritable state. At this time, after reading operations are processed, socket buffer may have free space for sending, but usually the parsed requests will be placed in new goroutines for processing. If the kernel cannot handle sending in time during this period, an invalid write system call will be wasted. Please close this pr at will. It seems that the upgrade will not be too obvious, but will lead to an increase in cpu overhead.

**Comment by limpo1989** [comment]
@panjf2000 这个pr没有从根源上修复问题，只是从表面上优化了发送效率（读一次之后如果有可写就强制发送一次），真正的根源是epoll事件机制出现了问题了

modWrite 通常会包含 read | write 实际上这样干会有两个问题：
1. 一段时间后（约10s之后）read事件触发次数比write触发次数多的多（读多写少）
2. 如果恶意客户端只发不收将会导致服务进程oom（实际上不需要恶意客户端，tcpkali就能稳定复现oom问题）

gnet echo 测试可以稳定复现oom （服务总内存128G，gnet压测30s左右内存峰值110G后几秒被oom）： 


image

```
[1615784.904275] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=init.scope,mems_allowed=0-1,global_oom,task_memcg=/user.slice/user-1000.slice/session-783.scope,task=echo-gnet-serve,pid=1477890,uid=1000
[1615784.904335] Out of memory: Killed process 1477890 (echo-gnet-serve) total-vm:122038268kB, anon-rss:118361296kB, file-rss:2304kB, shmem-rss:0kB, UID:1000 pgtables:232148kB oom_score_adj:0
[1615789.840989] oom_reaper: reaped process 1477890 (echo-gnet-serve), now anon-rss:0kB, file-rss:2304kB, shmem-rss:0kB
```

关于这个问题如果要修复应该就是 modWrite时去掉 read事件注册，只保留 write | error 事件，同理 modRead时只保留 read | error 注册

另外：evio 没有这个问题的原因时它不按事件类型进行处理，而是按照自己的状态决定（读优先级最低）：

https://github.com/tidwall/evio/blob/master/evio_unix.go#L229-L246
```go
//fmt.Println("-- loop started --", l.idx)
l.poll.Wait(func(fd int, note interface{}) error {
  if fd == 0 {
	  return loopNote(s, l, note)
  }
  c := l.fdconns[fd]
  switch {
  case c == nil:
	  return loopAccept(s, l, fd)
  case !c.opened:
	  return loopOpened(s, l, c)
  case len(c.out) > 0:
	  return loopWrite(s, l, c) // 如果有可写优先处理可写
  case c.action != None:
	  return loopAction(s, l, c)
  default:
	  return loopRead(s, l, c) // 读取操作优先级最低
  }
})
```

**Comment by panjf2000** [comment]
基于 I/O 多路复用的网络库的模式大多都是基于系统给的事件类型来做相应处理的。至于你说的这种发送快接收慢的客户端，只要是像 `gnet` 这样把业务层待写的数据保存起来，那就肯定都会有这个问题，即便是不替业务层保存待写的数据，那业务层就要自己保存这些数据，一样会有这个问题，这个也算是这一类网络框架经典的写出数据堆积问题了，性价比最高的一种解决办法是服务端设置一个阀值，业务层每次回写数据的时候都检查一下当前的待写数据堆积是否已经触达这个阀值，为了避免 OOM 通常需要服务端丢弃数据，`netty` 就是使用的这种方式，它提供了一个 `setWriteBufferHighWaterMark` 方法。至于 `gnet` 实现起来就更简单了：目前已经有了 `OutboundBuffered()` 可以随时检查当前的写出队列的堆积情况了，业务层自己就可以判断了。 

至于 evio 的 I/O 处理，反而是非主流的做法，虽然确实能解决这个问题，但因为这种做法每次只处理一种单一的事件，无法充分利用 `epoll` 的事件通知机制，性能上肯定有所损耗。还有你提出的解决的方法，虽然理论上也能解决这个问题，但是成本不小，一旦那么做了，后续还需要再重新把可读事件加回去，而这需要通过一个系统调用完成，在 LT 模式下待写队列大概率经常会有一些堆积，但是通常能很快被客户端读走，所以绝大部分时候不会出现 OOM。如果使用你说的方式，则需要频繁地删除然后重新添加可读事件，性能肯定下降，这相当于是将某一个极端场景的修复成本强加到大部分正常场景上，这是很不划算的。

还有，对于现实世界来说，性能问题的解决往往不能都丢给服务端，这样只会让服务端的实现越来越复杂，为了覆盖每一个业务问题而导致代码越来膨胀，这通常也是一个不好的实践。拿这个客户端发送快读取慢的问题来说，除了服务端可以使用上述的方法缓解之外，更重要的是对客户端进行改造，平衡好生产和消费的速率，毕竟一个疯狂发送数据而不接收处理数据的客户端有什么实际价值呢？在我看来这种客户端只有一种功能：做性能压测探一探服务端的极限边界，除此之外并没有现实的业务价值。而为了一个没有实际应用价值的极端场景去做一个会降低性能的妥协方案，则更没有价值了。

**Comment by gh-translator** [comment]
> 🤖 Non-English text detected, translating...         
----    
Most of the modes of network libraries based on I/O multiplexing are based on the event types given by the system for corresponding processing. As for the client you mentioned that sends quickly and receives slowly, as long as it saves the data to be written in the business layer like `gnet`, it will definitely have this problem, even if it does not save the data to be written for the business layer. , then the business layer will have to save the data itself, which will still have this problem. This is also a classic problem of data accumulation in this type of network framework. The most cost-effective solution is to set a threshold on the server side, and the business Every time the layer writes back data, it checks whether the current accumulation of data to be written has reached this threshold. In order to avoid OOM, the server usually needs to discard the data. `netty` uses this method. It provides a `setWriteBufferHighWaterMark ` method. As for `gnet`, it is even simpler to implement: there is already `OutboundBuffered()` that can check the current accumulation of the write queue at any time, and the business layer can judge it by itself. 

As for evio's I/O processing, it is a non-mainstream approach. Although it can indeed solve this problem, because this approach only handles a single event at a time, it cannot make full use of the event notification mechanism of `epoll`. In terms of performance, There must be some loss. There is also the solution you proposed. Although it can theoretically solve this problem, the cost is not small. Once it is done, the readable events need to be added back later, and this needs to be completed through a system call. In LT mode, there is a high probability that there will be some accumulation in the queue to be written, but it can usually be read by the client quickly, so OOM will not occur most of the time. If you use the method you mentioned, you need to frequently delete and re-add readable events, and the performance will definitely decrease. This is equivalent to imposing the repair cost of an extreme scenario on most normal scenarios, which is not cost-effective.

Also, in the real world, the solution to performance problems often cannot be left to the server. This will only make the implementation of the server more and more complex. In order to cover every business problem, the code will become more and more bloated. This is usually also a problem. Bad practice. Take the client's problem of fast sending and slow reading as an example. In addition to the above methods that the server can use to alleviate the problem, it is more important to transform the client to balance the production and consumption rates. After all, a person who sends data crazily cannot What real value is there in a client that doesn't receive and process data? In my opinion, this kind of client has only one function: to do performance stress testing to explore the limits of the server. Other than that, it has no real business value. It is even less valuable to make a compromise that will reduce performance for an extreme scenario that has no practical application value.

**Comment by limpo1989** [comment]
这种方式确实有损性能，毕竟不能同时处理读/写了，但是能充分利用tcp的滑动窗口限制对端的发送能力

按照当前的实现，echo 测试已经没办法进行了，基本上都会oom，如果加上丢弃或者断开连接那样也就脱离了echo测试的场景

就看如何取舍了，性能 / 安全

**Pull request closed without merging**